File: backend/src/main/java/com/joying/chat/service/RedisPubSubPublisher.java
package com.joying.chat.service;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.data.redis.core.RedisTemplate;
import org.springframework.stereotype.Service;

import com.joying.chat.config.RedisPubSubConfig;
import com.joying.chat.dto.ChatMessageResponse;

import lombok.RequiredArgsConstructor;

/**
 * 채팅 메시지를 Redis로 내보낸다.
 *
 * <p>서버가 여러 대일 때 자기에게 붙어 있지 않은 상대에게도 닿아야 하므로, 로컬
 * 브로커로 바로 보내지 않고 여기를 거친다.
 */
@Service
@RequiredArgsConstructor
public class RedisPubSubPublisher {

	private static final Logger log = LoggerFactory.getLogger(RedisPubSubPublisher.class);

	private final RedisTemplate<String, ChatMessageResponse> chatMessageRedisTemplate;

	/**
	 * 발행에 실패하면 예외를 올린다.
	 *
	 * <p>메시지는 저장까지 끝난 뒤에 발행되므로, 여기서 실패하면 저장은 됐는데 상대가
	 * 못 받은 상태가 된다. 부르는 쪽이 그것을 알아야 한다.
	 */
	public void publish(ChatMessageResponse message) {
		try {
			chatMessageRedisTemplate.convertAndSend(
				RedisPubSubConfig.CHAT_MESSAGE_CHANNEL, message);

			log.debug("Redis Pub/Sub 발행 성공: chatRoomId={}, senderId={}, messageId={}",
				message.getChatRoomId(), message.getSenderId(), message.getId());
		} catch (Exception e) {
			log.error("Redis Pub/Sub 발행 실패: chatRoomId={}, senderId={}, error={}",
				message.getChatRoomId(), message.getSenderId(), e.getMessage(), e);
			throw new RuntimeException("메시지 발행 실패", e);
		}
	}
}

File: backend/src/main/kotlin/com/joying/chat/service/RedisPubSubPublisher.kt
package com.joying.chat.service

import com.joying.chat.config.RedisPubSubConfig
import com.joying.chat.dto.ChatMessageDto
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.slf4j.LoggerFactory
import org.springframework.data.redis.core.RedisTemplate
import org.springframework.stereotype.Service

/**
 * Redis Pub/Sub Publisher
 *
 * 채팅 메시지를 Redis Pub/Sub로 발행
 * - Redis Stream 대신 간단한 Pub/Sub 사용
 * - 코루틴 suspend 함수로 비동기 처리
 * - Scale-out 가능 (여러 서버 인스턴스)
 */
@Service
class RedisPubSubPublisher(
    private val chatMessageRedisTemplate: RedisTemplate<String, ChatMessageDto>
) {
    private val logger = LoggerFactory.getLogger(RedisPubSubPublisher::class.java)

    /**
     * 채팅 메시지를 Redis Pub/Sub로 발행
     *
     * @param message 발행할 메시지
     */
    suspend fun publish(message: ChatMessageDto) {
        withContext(Dispatchers.IO) {
            try {
                // Redis Pub/Sub로 발행 (PUBLISH 명령)
                chatMessageRedisTemplate.convertAndSend(
                    RedisPubSubConfig.CHAT_MESSAGE_CHANNEL,
                    message
                )

                logger.debug(
                    "Redis Pub/Sub 발행 성공: chatRoomId={}, senderId={}, messageId={}",
                    message.chatRoomId,
                    message.senderId,
                    message.id
                )
            } catch (e: Exception) {
                logger.error(
                    "Redis Pub/Sub 발행 실패: chatRoomId={}, senderId={}, error={}",
                    message.chatRoomId,
                    message.senderId,
                    e.message,
                    e
                )
                throw RuntimeException("메시지 발행 실패", e)
            }
        }
    }
}